tool/src/org/antlr/v4/codegen/Target.java

```diff
@@ -92,14 +88,22 @@ public String getVersion() {
 	}
 
 	public STGroup getTemplates() {
+		String language = getLanguage();
+		STGroup templates = languageTemplates.get(language);
+
 		if (templates == null) {
```

Comment:
We might have a race here. If one thread gets templates==null it will enter the IF. Another thread could also get null and enter the IF before the synchronization occurs. Maybe we should move the synchronization under the whole function or surround function body with `synchronized (languageTemplates)`?

Reply:
I don't think it's a race here. In the code below `synhronized` is used:

```java
if (templates == null) {
	synchronized (languageTemplates) {
		templates = languageTemplates.get(language);
		if (templates == null) {
```

I use double-check for the best performance, locking is used only one time. It's a commonly used pattern: https://en.wikipedia.org/wiki/Double-checked_locking

Plain reading of the field is thread-safe.

Reply:
Ok, it looks like it's better to use `ConcurrentHashMap` in any case of multithreading: https://stackoverflow.com/q/33350489/1046374

Reply:
cool.

Reply:
It looks like the performance hit is insignificant: https://stackoverflow.com/a/12008833/1046374

Since `synchonized` on method looks clearer maybe it makes sense to use it everywhere.

Reply:
yeah, much simpler to do on method level.